addons/account/models/account_payment.py

```diff
@@ -168,6 +170,22 @@ class AccountPayment(models.Model):
         ),
     ]
 
+    def init(self):
+        super().init()
+        create_index(
+            self.env.cr,
+            indexname='account_payment_journal_id_company_id_idx',
+            tablename='account_payment',
+            expressions=['journal_id', 'company_id']
+        )
+        create_index(
+            self.env.cr,
+            indexname='account_payment_unmatched_idx',
+            tablename='account_payment',
+            expressions=['journal_id', 'company_id'],
+            where="NOT is_matched OR is_matched IS NULL"
+        )
```

Comment:
The partial index is small and over the life of a database shouldn't grow too much, because the payment sooner or later is matched, removing it from the keys of the index.
It would slightly slow down INSERTs and non-HOT UPDATE of unmatched payments, which shouldn't be much.

As for the full compound index, `journal_id` and `company_id` usually aren't UPDATEd too much once a payment is created, so only at the initial INSERT will there be some write overhead due to the index.
So in most cases, eventual overhead of the index is just tuple reference updating for non-HOT UPDATEs, which is best-case scenario for an index write overhead.
And as compound index, it replaces potentially 2 single indexes, on `journal_id` and `company_id`, which is nice.